File: backend/javascript/src/services/ingestion/ingestion.service.ts
import { Injectable, Logger } from "@nestjs/common";
import { FileSystemLoader } from "./loaders/file-system.loader";
import { MetadataBuilder } from "./processors/metadata.builder";
import { Chunker } from "./processors/chunker";
import { EmbeddingBuilder } from "./processors/embedding.builder";
import { VectorStoreManager } from "../providers/vectorstore/vector-store-manager";

/**
 * [인제션 오케스트레이터 서비스]
 * 
 * 로드 → 분류 → 분할 → 임베딩 → 저장의 전체 '인제션 파이프라인'을
 * 총괄 지휘하는 메인 서비스 클래스입니다.
 */
@Injectable()
export class IngestionService {
    private readonly logger = new Logger(IngestionService.name);

    constructor(
        private readonly loader: FileSystemLoader,
        private readonly metadataBuilder: MetadataBuilder,
        private readonly chunker: Chunker,
        private readonly embeddingBuilder: EmbeddingBuilder,
        private readonly vectorStore: VectorStoreManager,
    ) { }

    /**
     * 지정된 경로의 모든 문서를 처리하여 벡터 저장소에 쌓습니다.
     * 
     * @param docsRoot - 처리할 문서들이 담긴 최상위 폴더 경로
     */
    async run(docsRoot: string): Promise<void> {
        this.logger.log(`>>> Ingestion Started: [${docsRoot}]`);

        // 1. 문서 로드 (FileSystem I/O)
        const rawDocs = await this.loader.loadAllDocs(docsRoot);
        this.logger.log(`Step 1: Found ${rawDocs.length} files to process.`);

        // 파일 병렬 처리
        // 너무 많은 파일 동시에 처리 시 API 할당량 초과(Rate Limit) 날 수 있음
        // Promise.all을 사용해 모든 작업을 동시에 던짐( 파일이 수천 개라면 10개씩 끊어서 처리하는 'Batching' 로직 추가 필요 )
        const processResults = await Promise.all(rawDocs.map(async (doc) => {
            try {
                // 2-1. 메타데이터 생성 (AI 분석 포함)
                const meta = await this.metadataBuilder.build(doc.filePath, doc.rawText);
                // 2-2. 청킹 (의미 기반 텍스트 분할)
                const chunks = await this.chunker.splitDocument(doc.rawText, meta);
                // 2-3. 임베딩 (AI 좌표 추출)
                const embedded = await this.embeddingBuilder.embed(chunks);
                // 2-4. 벡터 저장소 저장
                await this.vectorStore.storeChunks(embedded);
                return chunks.length; // 성공한 청크 수 반환
            } catch (error: any) {
                this.logger.error(`Failed to process ${doc.filePath}: ${error.message}`);
                return 0; // 실패 시 0개
            }
        }));
        // 총 처리된 청크 수 합산
        const totalChunksCount = processResults.reduce((acc, count) => acc + count, 0);
        this.logger.log(`>>> Ingestion Finished. Total: ${rawDocs.length} files / ${totalChunksCount} chunks processed.`);
    }

    /**
     * 개별 파일을 처리하여 벡터 저장소에 추가
     * 
     * @param filename - 원본 파일명
     * @param content - 파일 내용
     * @param filePath - 파일 경로 (메타데이터용)
     * @returns 처리 결과 정보
     */
    async processFile(filename: string, content: string, filePath: string): Promise<{
        filename: string;
        filePath: string;
        chunkCount: number;
        title: string;
        docType: string;
        updatedAt: string;
    }> {
        this.logger.log(`>>> Processing single file: [${filename}]`);

        try {
            // 1. 메타데이터 생성 (AI 분석 포함)
            const meta = await this.metadataBuilder.build(filePath, content);
            this.logger.log(`Step 1: Metadata built for ${filename}`);

            // 2. 청킹 (의미 기반 텍스트 분할)
            const chunks = await this.chunker.splitDocument(content, meta);
            this.logger.log(`Step 2: Split into ${chunks.length} chunks`);

            // 3. 임베딩 (AI 좌표 추출)
            const embedded = await this.embeddingBuilder.embed(chunks);
            this.logger.log(`Step 3: Embedded ${embedded.length} chunks`);

            // 4. 벡터 저장소 저장
            await this.vectorStore.storeChunks(embedded);
            this.logger.log(`Step 4: Stored ${embedded.length} chunks to vector store`);

            this.logger.log(`>>> File processing completed: ${filename} (${chunks.length} chunks)`);

            // 처리 결과 반환
            return {
                filename,
                filePath,
                chunkCount: chunks.length,
                title: meta.title,
                docType: meta.docType,
                updatedAt: meta.updatedAt
            };
        } catch (error: any) {
            this.logger.error(`Failed to process ${filename}: ${error.message}`);
            throw error; // 에러를 상위로 전파하여 컨트롤러에서 처리
        }
    }
}

File: backend/java/src/main/java/com/ragassistant/service/IngestionService.java
package com.ragassistant.service;

import com.ragassistant.model.DocumentChunk;
import com.ragassistant.provider.chunking.MarkdownHeaderChunker;
import com.ragassistant.provider.embedding.EmbeddingBuilder;
import com.ragassistant.provider.metadata.MetadataProvider;
import com.ragassistant.provider.vectorstore.ChromaVectorStore;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import com.ragassistant.model.SourceDocument;
import com.ragassistant.repository.SourceDocumentRepository;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;

@Slf4j
@Service
@RequiredArgsConstructor
public class IngestionService {
    private final MetadataProvider metadataProvider;
    private final MarkdownHeaderChunker chunker;
    private final EmbeddingBuilder embeddingBuilder;
    private final ChromaVectorStore vectorStore;
    private final SourceDocumentRepository sourceRepo;

    public ProcessResult processFile(String filename, String content, String filePath) {
        log.info(">>> Processing single file: [{}]", filename);

        try {
            // 1. Metadata extraction
            DocumentChunk.ChunkMetadata meta = metadataProvider.extract(filePath, content, filename);
            log.info("Step 1: Metadata built for {}", filename);

            // 2. Chunking
            List<DocumentChunk> chunks = chunker.splitDocument(content, meta);
            log.info("Step 2: Split into {} chunks", chunks.size());

            // 3. Embedding
            List<DocumentChunk> embeddedChunks = embeddingBuilder.embed(chunks);
            log.info("Step 3: Embedded {} chunks", embeddedChunks.size());

            // 4. Vector Store
            vectorStore.storeChunks(embeddedChunks);
            log.info("Step 4: Stored {} chunks to vector store", embeddedChunks.size());

            // 5. SQLite Tracking
            SourceDocument doc = sourceRepo.findByFilePath(filePath)
                    .orElse(SourceDocument.builder().filePath(filePath).build());
            doc.setFilename(filename);
            doc.setTitle(meta.getTitle());
            doc.setDocType(meta.getDocType());
            doc.setVersion(meta.getVersion());
            doc.setChunkCount(chunks.size());
            doc.setUpdatedAt(meta.getUpdatedAt());
            doc.setCreatedAt(LocalDateTime.now());
            sourceRepo.save(doc);

            return ProcessResult.builder()
                    .filename(filename)
                    .filePath(filePath)
                    .chunkCount(chunks.size())
                    .title(meta.getTitle())
                    .docType(meta.getDocType())
                    .updatedAt(meta.getUpdatedAt())
                    .build();
        } catch (Exception e) {
            log.error("Failed to process {}: {}", filename, e.getMessage());
            throw new RuntimeException(e);
        }
    }

    public static class ProcessResult {
        private String filename;
        private String filePath;
        private int chunkCount;
        private String title;
        private String docType;
        private String updatedAt;

        public ProcessResult() {
        }

        public static ProcessResultBuilder builder() {
            return new ProcessResultBuilder();
        }

        public String getFilename() {
            return filename;
        }

        public void setFilename(String filename) {
            this.filename = filename;
        }

        public String getFilePath() {
            return filePath;
        }

        public void setFilePath(String filePath) {
            this.filePath = filePath;
        }

        public int getChunkCount() {
            return chunkCount;
        }

        public void setChunkCount(int chunkCount) {
            this.chunkCount = chunkCount;
        }

        public String getTitle() {
            return title;
        }

        public void setTitle(String title) {
            this.title = title;
        }

        public String getDocType() {
            return docType;
        }

        public void setDocType(String docType) {
            this.docType = docType;
        }

        public String getUpdatedAt() {
            return updatedAt;
        }

        public void setUpdatedAt(String updatedAt) {
            this.updatedAt = updatedAt;
        }

        public static class ProcessResultBuilder {
            private String filename;
            private String filePath;
            private int chunkCount;
            private String title;
            private String docType;
            private String updatedAt;

            public ProcessResultBuilder filename(String f) {
                this.filename = f;
                return this;
            }

            public ProcessResultBuilder filePath(String f) {
                this.filePath = f;
                return this;
            }

            public ProcessResultBuilder chunkCount(int c) {
                this.chunkCount = c;
                return this;
            }

            public ProcessResultBuilder title(String t) {
                this.title = t;
                return this;
            }

            public ProcessResultBuilder docType(String d) {
                this.docType = d;
                return this;
            }

            public ProcessResultBuilder updatedAt(String u) {
                this.updatedAt = u;
                return this;
            }

            public ProcessResult build() {
                ProcessResult r = new ProcessResult();
                r.setFilename(filename);
                r.setFilePath(filePath);
                r.setChunkCount(chunkCount);
                r.setTitle(title);
                r.setDocType(docType);
                r.setUpdatedAt(updatedAt);
                return r;
            }
        }
    }

    public List<SourceDocument> getAllSources() {
        return sourceRepo.findAll();
    }

    public void deleteAllSources() {
        sourceRepo.deleteAll();
        // TODO: Also delete from vector store if needed
        log.info("Deleted all source documents from database");
    }
}
